### PR TITLE
KAFKA-10500: Add failed-stream-threads metric for adding + removing stream threads

### PR DESCRIPTION
**Pull request by lct45**
Per [KIP-663](https://cwiki.apache.org/confluence/display/KAFKA/KIP-663%3A+API+to+Start+and+Shut+Down+Stream+Threads), adding a metric to record the failed streams threads over the life of a client.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by lct45** [comment]
@wcarlson5 @cadonna for initial review

File: streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java
Comment (lct45) [inline streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java]: We also need to verify that the metric works when there is a failed stream thread. Options are (1) to create a custom processor now and (IIRC) run the test suite twice, once with failing stream threads and once without to confirm that the metric works. I'm not sure if the custom processor will let us just fail one stream thread right before closing the app. Or (2) wait until add/remove stream threads is implemented and remove threads and test the metric after removing some threads before closing the app. WDYT?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: This seems wrong. There are a few duplicate calls here

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: We probably don't want to skip this log. Can you move the sensor in here?

**Review by wcarlson5 — COMMENTED** [review]
I am not sure it is necessary but you might want to add an integration test that kills a few threads and check the metrics. You would need to set the old handler to get a single thread to die as of now

File: streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java]: I think you should try using the custom processor. You can find an example in `StreamsUncaughtExceptionHandlerIntegrationTest.java`

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (lct45) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Ahh, got mixed up when I rebased trunk

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Could you please add a public method to `StreamsMetricsImpl` named `removeAllClientLevelSensorsAndMetrics()` that calls `removeAllClientLevelMetrics()` and `removeAllClientLevelSensors()` and make the latter two methods `private`?

File: streams/src/main/java/org/apache/kafka/streams/internals/metrics/ClientMetrics.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/internals/metrics/ClientMetrics.java]:
```suggestion
    private static final String FAILED_STREAM_THREADS_DESCRIPTION = "The number of failed stream threads since the start of the Kafka Streams client";
```

File: streams/src/test/java/org/apache/kafka/streams/internals/metrics/ClientMetricsTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/internals/metrics/ClientMetricsTest.java:140]: nit: I like to insert a blank line after the call to test to visually separate setup, call, and verification.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:218]: Unit tests for this method are missing.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: Unit tests for this method are missing. Please also consider my comment in class `KafkaStreams` for these unit tests.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: Here you should just need a queue as for `clientLevelMetrics`. We need a map for the other levels because there can be multiple objects for each level, e.g., there might be multiple stream thread and each one manages its sensors under a key in the map. However, there is only one client on client level.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Not every dead stream thread is a failed stream thread. You should record this metric where the uncaught exception handler is called because there we now that a stream thread died unexpectedly.

File: streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java]: I would put the test whether the metric is recorded correctly in `StreamThreadTest`. An example for such a test is `shouldLogAndRecordSkippedRecordsForInvalidTimestamps()`. I do not think an integration test is needed. The test regarding the existence of the metric, i.e., `checkMetricByName(listMetricThread, FAILED_STREAM_THREADS, 1);` should stay here.

**Review by cadonna — COMMENTED** [review]
@lct45 Thank you for the PR!

Here my feedback.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (lct45) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Would that just be in `run()` of the GlobalStreamThread then?

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: No, that would be in `StreamThread#runLoop()`.

File: streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java
Comment (lct45) [reply streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java]: After looking at both test classes, I think it actually might make the most sense to put the test for this metric in `StreamsUncaughtExceptionHandlerIntegrationTest`, since the metric is so closely aligned with the exception handler anyways and the setup works nicely with what we're trying to test with the metric. From the size + complexity of the other test classes, I think creating an overloaded processor for one test out of 20+ tests seems tricky.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: You can inline the value of variable `key` here and remove `key`.

```suggestion
            final String fullSensorName = CLIENT_LEVEL_GROUP + SENSOR_NAME_DELIMITER + sensorName;
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:227]: Although, we use this in other methods, I think the following is a bit simpler to read:
```suggestion
        final Sensor sensor = metrics.getSensor(fullSensorName);
        if (sensor == null) {
            clientLevelSensors.push(fullSensorName);
            return metrics.sensor(fullSensorName, recordingLevel, parents);
        }
        return sensor;
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java]:
```suggestion
        final Sensor actualSensor = streamsMetrics.clientLevelSensor(SENSOR_NAME_1, recordingLevel);
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java]:
```suggestion
        final Sensor actualSensor = streamsMetrics.clientLevelSensor(SENSOR_NAME_1, recordingLevel);
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java]:
```suggestion
        metrics.removeSensor(sensorKeys.getValues().get(0));
        metrics.removeSensor(sensorKeys.getValues().get(1));
        expect(metrics.removeMetric(metricName1)).andReturn(mock(KafkaMetric.class));
        expect(metrics.removeMetric(metricName2)).andReturn(mock(KafkaMetric.class));
        replay(metrics);

        streamsMetrics.removeAllClientLevelSensorsAndMetrics();

        verify(metrics);
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java:664]: Please remove this empty line.

File: streams/src/test/java/org/apache/kafka/streams/internals/metrics/ClientMetricsTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/internals/metrics/ClientMetricsTest.java]: Please remove empty line.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: You can remove these lines. They are dead code.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]:
```suggestion
        EasyMock.replay(taskManager);
        thread.updateThreadMetadata("metadata");
        thread.setState(StreamThread.State.STARTING);
        
        thread.runLoop();

        final Metric failedThreads = StreamsTestUtils.getMetricByName(metrics.metrics(), "failed-stream-threads", "stream-metrics");
        assertThat(failedThreads.metricValue(), is(shouldFail ? 1.0 : 0.0));
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: I would rename this method to `runAndVerifyFailedStreamThreadRecording()`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Could you please specify two separate unit tests? One could be named `shouldRecordFailedStreamThread()` and the other `shouldNotRecordFailedStreamThread()`.

**Review by cadonna — COMMENTED** [review]
@lct45 Thank you for the updates!

I have rather minor comments.

**Review by cadonna — APPROVED** [review]
LGTM!

Call for committer review and merge:  @ableegoldman @mjsax @vvcephei @guozhangwang @abbccdda

File: streams/src/main/java/org/apache/kafka/streams/internals/metrics/ClientMetrics.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/internals/metrics/ClientMetrics.java:135]: nit: missing empty line

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: Should we rewrite this the same way `threadLevelSensor` is written (ie, using `orElseGet`) for consistency?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java]: Why did we change this from `andStubReturn(null)` to `andReturn(mock(KafkaMetric.class))`?

**Review by mjsax — COMMENTED** [review]
Just some nits

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: I requested this. See my comment https://github.com/apache/kafka/pull/9614#discussion_r529313022

Comment (lct45) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: I'm good either way (:

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (lct45) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java]: Must've been an accidental change when trying to get the test to work. `shouldRemoveStateStoreLevelSensors` uses `andReturn(mock(KafkaMetric.class))` so that's where it came from, but this test works with `andStubReturn(null)` so I changed it back to that

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (lct45) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java:623]: This wasn't being used so I went ahead and took it out

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: I am fine either way, too, but I prefer consistency... So should we rewrite the other method as a side cleanup?

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: I am fine with consistency and clean-up, but I would like to have the clean-up in a separate PR.

Comment (lct45) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: I changed it back for consistency and will open up a fix PR to update both of them to the new syntax

**Comment by mjsax** [comment]
Retest this please

**Review by mjsax — APPROVED** [review]
LGTM. Will merge after the build passed.

**Pull request merged**
